Add make_timestamp sparksql function

marin-ma opened the pull request:
Add sparksql function `make_timestamp`, which creates a timestamp from year, month, day, hour, min, sec and timezone (optional) fields. The timezone field indicates the timezone of the input timestamp. If specified, the output timestamp will be converted to the local session timezone.

```
set spark.sql.session.timeZone=Asia/Shanghai;
SELECT make_timestamp(2014, 12, 28, 6, 30, 45.887); -- 2014-12-28 06:30:45.887
SELECT make_timestamp(2014, 12, 28, 6, 30, 45.887, 'CET'); -- 2014-12-28 13:30:45.887
```

Spark documentation:

https://spark.apache.org/docs/latest/api/sql/#make_timestamp

Spark implementation:

https://github.com/apache/spark/blob/branch-3.3/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L2512-L2712

marin-ma commented:
cc: @rui-mo @PHILO-HE

rui-mo commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,161 @@
+      // use default value 0(UTC timezone).
+      int64_t sessionTzID = 0;
+      const auto& queryConfig = context.execCtx()->queryCtx()->queryConfig();
+      const auto sessionTzName = queryConfig.sessionTimezone();
```
By declaring this function as stateful, sessionTzName could become a construction parameter of MakeTimestampFunction, instead of being calculated for each apply.

rui-mo commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -757,5 +758,118 @@ TEST_F(DateTimeFunctionsTest, fromUnixtime) {
+    auto result = hasTimeZone
+        ? evaluate("make_timestamp(c0, c1, c2, c3, c4, c5, c6)", data)
+        : evaluate("make_timestamp(c0, c1, c2, c3, c4, c5)", data);
+    facebook::velox::test::assertEqualVectors(expected, result);
```
Maybe we need to test different encodings as a fast path for constant encoding is implemented. `testEncodings` could be used for that purpose.

mbasmanova replied:
@rui-mo Rui has a good point. Would you address her comment?

marin-ma replied:
@mbasmanova Method `testConstantTimezone` below was used to address this comment.

marin-ma commented:
@rui-mo Could you help to review again? Thanks!

PHILO-HE commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,199 @@
+    auto localMicros =
+        hour * util::kMicrosPerHour + minute * util::kMicrosPerMinute + micros;
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxException& e) {
```
If user error can only be `VeloxUserError` here, maybe we can just catch such exception.

PHILO-HE commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,199 @@
+      throw;
+    }
+    return std::nullopt;
+  } catch (const std::exception&) {
```
Can be removed?

PHILO-HE commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,199 @@
+        monthVector->valueAt<int32_t>(row),
+        dayVector->valueAt<int32_t>(row));
+    auto localMicros =
+        hour * util::kMicrosPerHour + minute * util::kMicrosPerMinute + micros;
```
Can we simply call `util::fromTime()`?

PHILO-HE commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,199 @@
+      }
+    } else {
+      // Otherwise use session timezone. If session timezone is not specified,
+      // use default value 0(UTC timezone).
```
I once noted Spark always has a session timezone in its config. By default, it's the one detected from OS. And we always let Gluten pass session timezone to Velox. So maybe, if session timezone is not found from config, we can simply throw an exception.

PHILO-HE reviewed (commented):
Let's keep consistency for PR title by using "Add xxx Spark function". Thanks!
See community convention: https://github.com/facebookincubator/velox/blob/main/CONTRIBUTING.md#adding-scalar-functions

rui-mo commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+  VELOX_USER_CHECK(
+      inputArgs[5].type->isShortDecimal(),
+      "Seconds must be short decimal type but got {}",
+      inputArgs[5].type->toString());
```
Maybe add type checks for all input args.

marin-ma replied:
I wonder why it would be necessary to check all input args. This decimal check is aligned with Spark's https://github.com/apache/spark/blob/e98872fb5d07d570e6d0516b49a5d2e58876d1a6/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala#L2601-L2602

rui-mo commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+    auto localMicros = util::fromTime(hour, minute, 0, (int32_t)micros);
+    return util::fromDatetime(daysSinceEpoch, localMicros);
+  } catch (const VeloxUserError& e) {
+    return std::nullopt;
```
To use try-catch for each row may have poor performace. Maybe we can take use of Status to represent the computing outcome of `daysSinceEpochFromDate` method.

marin-ma replied:
Maybe we can optimize that in a separate PR. @mbasmanova Do you have any suggestion?

mbasmanova replied:
@rui-mo Rui has a good point. We should not throw and catch exceptions per row. See https://velox-lib.io/blog/optimize-try_cast.

marin-ma commented:
@mbasmanova Could you help to review? Thanks!

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.h:
```diff
@@ -14,8 +14,10 @@
  * limitations under the License.
  */
 
+#include <boost/algorithm/string.hpp>
```
Are these includes needed?

marin-ma replied:
CI Compilation would fail without this include:

```
2024-03-06T15:38:41.3214677Z       In file included from /project/velox/functions/sparksql/DateTimeFunctions.cpp:17:
2024-03-06T15:38:41.3215875Z       /project/./velox/functions/sparksql/DateTimeFunctions.h: In static member function ‘static std::optional<signed char> facebook::velox::functions::sparksql::NextDayFunction<T>::getDayOfWeekFromString(const facebook::velox::StringView&)’:
2024-03-06T15:38:41.3216354Z       /project/./velox/functions/sparksql/DateTimeFunctions.h:583:9: error: ‘boost’ has not been declared
2024-03-06T15:38:41.3216558Z         583 |         boost::algorithm::to_lower_copy(dayOfWeek.str());
2024-03-06T15:38:41.3216661Z             |         ^~~~~
```

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+  }
+}
+
+class MakeTimestampFunction : public exec::VectorFunction {
```
Can this function be implemented as a simple function?

marin-ma replied:
The 6th parameter is of decimal type. I'm not sure it it's possible to implement it as a simple function.

mbasmanova replied:
> The 6th parameter is of decimal type.

Wow... I didn't realize that. Would you update the documentation to state that clearly? Are there restrictions on precision and scale? I assume scale must be <= 3.

mbasmanova replied:
@marin-ma Can you remind me why can't we implement functions with decimal inputs as simple functions? I wonder if it is worth looking into extending the framework to support such functions. Does Fuzzer support such functions? If not, we need to extend it. 

CC: @rui-mo @PHILO-HE @FelixYBW

marin-ma replied:
@mbasmanova Looks like the `resolver` used by simple fucntions doesn't support resolving decimal type https://github.com/facebookincubator/velox/blob/main/velox/expression/UdfTypeResolver.h
The expression fuzzer test doesn't use above framework so it can create decimal vectors.

mbasmanova replied:
> The expression fuzzer test doesn't use above framework so it can create decimal vectors.

@marin-ma Are you saying that Fuzzer does cover this function? Would you run the Fuzzer with --only make_timestamp to make sure there are no failures?

marin-ma replied:
@mbasmanova I get exception with this command `./velox/expression/tests/spark_expression_fuzzer_test --only make_timestamp`

```
E0311 22:59:16.962633 2410940 Exceptions.h:69] Line: /home/sparkuser/github/oap-project/velox/velox/expression/tests/ExpressionFuzzer.cpp:1415, Function:fuzzReturnType, Expression: !signatures_.empty() No function signature available., Source: RUNTIME, ErrorCode: INVALID_STATE
terminate called after throwing an instance of 'facebook::velox::VeloxRuntimeError'
  what():  Exception: VeloxRuntimeError
Error Source: RUNTIME
Error Code: INVALID_STATE
Reason: No function signature available.
Retriable: False
Expression: !signatures_.empty()
Function: fuzzReturnType
File: /home/sparkuser/github/oap-project/velox/velox/expression/tests/ExpressionFuzzer.cpp
```

Does it mean this function is not covered by the fuzzer test? Or did I use the wrong command?

mbasmanova replied:
@marin-ma I believe Fuzzer doesn't support DECIMAL types yet. It would be nice to add this support, otherwise, test coverage of VectorFunctions that use DECIMAL types is limited.

CC: @rui-mo @PHILO-HE @majetideepak

rui-mo replied:
@mbasmanova Yes, fuzzer test for decimal type is not supported. This comment is about my previous finding https://github.com/facebookincubator/velox/pull/5791#issuecomment-1842440425, and I will remove the two limitations to see where the gap is, thanks.

mbasmanova reviewed (commented):
Some comments.

marin-ma commented:
@mbasmanova Could you help to review again? Thanks!

marin-ma commented:
@mbasmanova Could you please help to review again? Thanks!

mbasmanova commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+  if (hour < 0 || hour > 24) {
+    return std::nullopt;
+  }
+  // Check miniute.
```
miniute - >minute

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -139,6 +139,35 @@ These functions support TIMESTAMP and DATE input types.
 
         SELECT quarter('2009-07-30'); -- 3
 
+.. spark:function:: make_timestamp(year, month, day, hour, min, sec[, timezone]) -> timestamp
```
do not abbreviate: min -> minute, sec -> second

mbasmanova commented on velox/docs/functions/spark/datetime.rst:
```diff
@@ -139,6 +139,35 @@ These functions support TIMESTAMP and DATE input types.
+    Otherwise the function assumes the inputs are in the session's configured time zone.
+    Requires ``session_timezone`` to be set, or an exceptions will be thrown.
+
+    Arguments:
```
Would you generate the docs and verify that they render nicely?

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -17,6 +17,7 @@
 #include "velox/common/base/tests/GTestUtils.h"
 #include "velox/functions/sparksql/tests/SparkFunctionBaseTest.h"
 #include "velox/type/tz/TimeZoneMap.h"
+#include "velox/vector/tests/utils/VectorTestBase.h"
```
Is this include needed?

mbasmanova commented on velox/functions/sparksql/tests/DateTimeFunctionsTest.cpp:
```diff
@@ -819,5 +820,147 @@ TEST_F(DateTimeFunctionsTest, fromUnixtime) {
+    testMakeTimestamp(data, expected, true);
+  }
+
+  // Invalid cases.
```
Would you split this test method into 2: valid and invalid cases.

mbasmanova reviewed (commented):
@marin-ma Overall looks good. Some comments below.

marin-ma commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+    VELOX_DCHECK(status.isUserError());
+    return std::nullopt;
+  }
+  // micros has at most 8 digits (2 for seconds + 6 for microseconds),
```
Seconds should between [0-60]. Invalid inputs will fail on this check https://github.com/marin-ma/velox-oap/blob/0da4ffae84ccacdd8257b803f61caf70da18d5e9/velox/functions/sparksql/DateTimeFunctions.cpp#L47-L50

marin-ma commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+    return {
+        exec::FunctionSignatureBuilder()
+            .integerVariable("precision")
+            .integerVariable("scale")
```
It throws exception with this change:

```
C++ exception with description "Exception: VeloxUserError
Error Source: USER
Error Code: INVALID_ARGUMENT
Reason: Scalar function signature is not supported: make_timestamp(INTEGER, INTEGER, INTEGER, INTEGER, INTEGER, DECIMAL(16, 6)). Supported signatures: (integer,integer,integer,integer,integer,decimal(precision,6)) -> timestamp, (integer,integer,integer,integer,integer,decimal(precision,6),varchar) -> timestamp.
Retriable: False
Function: resolveScalarFunctionType
File: /home/sparkuser/github/oap-project/velox/velox/parse/TypeResolver.cpp
Line: 99
Stack trace:
Stack trace has been disabled. Use --velox_exception_user_stacktrace_enabled=true to enable it.
" thrown in the test body.
```

mbasmanova replied:
@marin-ma This sounds like a bug. Would you create a GitHub issue and help investigate and fix?

CC: @majetideepak

marin-ma replied:
@mbasmanova Filed an issue here https://github.com/facebookincubator/velox/issues/9029
I will investigate this next.

marin-ma commented:
@mbasmanova Here's the generated doc:
![image](https://github.com/facebookincubator/velox/assets/52736607/3ef33300-7e3e-4114-9855-e1aece38e739)

marin-ma commented on velox/functions/sparksql/DateTimeFunctions.cpp:
```diff
@@ -0,0 +1,198 @@
+        rows.applyToSelected([&](vector_size_t row) {
+          auto timestamp = makeTimeStampFromDecodedArgs(
+              row, year, month, day, hour, minute, micros);
+          setTimestampOrNull(row, timestamp, constantTzID, resultFlatVector);
```
This behavior aligns with spark's output under non-ansi mode. Spark returns NULL if for invalid inputs, such as month > 12, seconds > 60, etc. Added examples with invalid inputs in the document.

mbasmanova commented:
@marin-ma #9044 landed. Would you rebase this PR so it can specify fixed scale in the function signature?

marin-ma commented:
@mbasmanova I've rebased my branch and addressed it in the latest commit https://github.com/marin-ma/velox-oap/tree/make-timestamp However, it appears that these changes have not been synchronized with the PR. Do you have any idea on how to resolve it?

mbasmanova commented:
> I've rebased my branch and addressed it in the latest commit 

Thanks.

> However, it appears that these changes have not been synchronized with the PR. Do you have any idea on how to resolve it?

What do you mean by "have not been synchronized with the PR"? Would you elaborate a bit?

marin-ma commented:
> What do you mean by "have not been synchronized with the PR"? Would you elaborate a bit?

Earlier, the commit history in this PR wasn't updated, and it didn't trigger CI after the rebase. But it looks normal now.

mbasmanova commented:
@marin-ma I missed it during review, but the new function should go into a separate MakeTimestamp.cpp file and the test should go into MakeTimestampTest.cpp. I'll make these changes before landing. You can see a preview in #9060 . I created this PR to get CI signals before landing.

marin-ma commented:
> @marin-ma I missed it during review, but the new function should go into a separate MakeTimestamp.cpp file and the test should go into MakeTimestampTest.cpp. I'll make these changes before landing. You can see a preview in #9060 . I created this PR to get CI signals before landing.

@mbasmanova Thanks. Do I need to port these changes to this PR?

mbasmanova commented:
> Do I need to port these changes to this PR?

@marin-ma No. I just let you know so you are not surprised to see these changes when this PR lands.
